extensions/oplogPopulator/allocationStrategy/AllocationStrategy.js

```diff
@@ -4,21 +4,68 @@ class AllocationStrategy {
 
     /**
      * @constructor
-     * @param {Logger} logger logger object
+     * @param {Object} params params
+     * @param {ConnectorsManager} params.connectorsManager connector manager
+     * @param {OplogPopulatorMetrics} params.metricsHandler metrics handler
+     * @param {Logger} params.logger logger object
      */
-    constructor(logger) {
-        this._logger = logger;
+    constructor(params) {
+        this._logger = params.logger;
+        this._metricsHandler = params.metricsHandler;
+        // Stores connector assigned for each bucket
+        this._bucketsToConnectors = new Map();
+        // Stores buckets that should are removed from the connector
+        // but still in use
+        this._retainedBuckets = new Map();
+        this._connectorsManager = params.connectorsManager;
+
+        this._initConnectorToBucketMap();
+
+        this._connectorsManager.on('update', () => {
+            this._retainedBuckets.forEach(bucket => {
+                if (!this._bucketsToConnectors.has(bucket)) {
+                    this._retainedBuckets.delete(bucket);
+                }
+            });
+            this._retainedBuckets.clear();
+        });
```

Comment:
I don't think this covers all cases.

- When we have an "old" connector, listening to multiple buckets : it may be kept longer, while the bucket will have been removed from `_bucketsToConnectors` already
- The (multiple) connectors are not updated synchronously, so we cannot clear everything atomically

--> should probably have an event when each connector is actually updated, to clear the retained buckets map accordingly:

```suggestion
        this._connectorsManager.on('update', updatedConnector => {
            this._retainedBuckets.forEach((bucket, connector) => {
                if (updatedConnector === connector) {
                    this._retainedBuckets.delete(bucket);
                }
            });
        });
```